Reimplement experimental load-balancing tests in scripttest

joamaki opened the pull request:
This PR adds a "k8s" script command and reimplements the k8s integration test-suite in `pkg/loadbalancer/experimental` as script tests.

StateDB is bumped to v0.3.2 to fix the column ordering bug.

joamaki committed:
experimental: Convert nodeport to script test

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Convert dualstack to script test

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
k8s/client: k8s fake client script command

Add the "k8s" script command for interacting with the client-go fake clientset.
Example usage in txtar:

  k8s add foo.yaml bar.yaml
  k8s update foo2.yaml
  k8s delete foo2.yaml bar.yaml

  -- foo.yaml --
  apiVersion: v1
  ...

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
k8s/testutils: Relocate API resource definitions from test/controlplane

The API resource definitions used by test/controlplane for detecting
capabilities are usable in other contexts. Move them to the testutils
package.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Convert hostport to script test

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Switch to debug logging

As things are maturing, switch most of the logging from Info to Debug.
Remove the unused logger from Writer.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Convert clusterip to script test

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Convert source-ranges to script test

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Convert graceful termination to script test

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Convert multiport to script test

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Add pruning test-case

Test for checking that unexpected BPF map contents are pruned when tables are empty.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Remove k8s_test.go and cleanup dead code

Now that tests have been converted to scripttest drop the old test suite.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
vendor: Bump StateDB to v0.3.2

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki commented on pkg/loadbalancer/experimental/script_test.go:
```diff
@@ -0,0 +1,97 @@
+)
+
+func TestScript(t *testing.T) {
+	version.Force(testutils.DefaultVersion)
```
I had an earlier version that allowed doing `--k8s-force-version=1.23` in the `txtar` files, but it really wrecks havoc with parallel testing and the `version` package globals, so ended up removing the test I had for older k8s versions (e.g. Endpoints no EndpointSlices). For now I'm just forcing the version here. Will have to revisit the `pkg/k8s/version` global variable issue in a later PR instead.

ovidiutirla replied:
Let's create an issue and link it here maybe

joamaki replied:
https://github.com/cilium/cilium/issues/35537

joamaki commented on pkg/loadbalancer/experimental/script_test.go:
```diff
@@ -0,0 +1,97 @@
+		}, []string{}, "testdata/*.txtar")
+}
+
+func parseArgsLine(file string) []string {
```
Wonder if this should be made into a builtin feature of scripttest and the args passed to the engine constructor. Not in this PR though.

joamaki replied:
Done in https://github.com/cilium/hive/pull/22

joamaki commented on pkg/loadbalancer/experimental/frontend.go:
```diff
@@ -89,6 +89,7 @@ func (fe *Frontend) TableHeader() []string {
 		"PortName",
 		"Backends",
 		"Status",
+		"StatusKind", // FIXME: Figure out a better way to do this.
```
This was added so that the tests can just check whether the frontend is marked "Done" without having to deal with the non-deterministic timestamp in `Status.String()` ("Done (10ms ago"). I'm kind of inclined to just leave it this way even though you'll then see "Done (10ms ago)   Done" in the table output. Alternative would be to break the Status into multiple columns (StatusKind, StatusSince, StatusMessage)...

ovidiutirla replied:
could we just split the output by space in this case for example when checking the output? splitting the columns in StatusKind / StatusSince sounds reasonable to me tho.

joamaki replied:
Split this into "Status", "Since" and "Error"

ovidiutirla commented on pkg/loadbalancer/experimental/config.go:
```diff
@@ -11,9 +11,10 @@ import (
+	EnableExperimentalLB          bool          `mapstructure:"enable-experimental-lb"`
+	RetryBackoffMin               time.Duration `mapstructure:"lb-retry-backoff-min"`
+	RetryBackoffMax               time.Duration `mapstructure:"lb-retry-backoff-max"`
+	TestFaultInjectionProbability float32       `mapstructure:"lb-test-fault-injection"`
```
I'm quite against adding Configs only for testing purposes

ovidiutirla replied:
Not super familiar with the feature itself but would be nice if we could separate the testing code from the production code.

joamaki replied:
I’ll split up this into a ”TestConfig” and make it an optional dependency and will only provide it in tests.

ovidiutirla commented on pkg/k8s/client/cell.go:
```diff
@@ -545,6 +587,86 @@ func FakeClientBuilder() ClientBuilderFunc {
+				if err != nil {
+					return nil, fmt.Errorf("failed to read %s: %w", file, err)
+				}
+				obj, gvk, err := testutils.DecodeObjectGVK([]byte(b))
```
`[]byte` is redundant here 
```suggestion
				obj, gvk, err := testutils.DecodeObjectGVK(b)
```

ovidiutirla commented on pkg/loadbalancer/experimental/benchmark/benchmark.go:
```diff
@@ -346,3 +352,245 @@ func ServicesAndSlices(testSize int) (svcs []*slim_corev1.Service, epSlices []*k
+						EnableExperimentalLB:          true,
+						RetryBackoffMin:               time.Millisecond,
+						RetryBackoffMax:               time.Millisecond,
+						TestFaultInjectionProbability: 0.0,
```
should we use `failureProbability`?

joamaki replied:
Yep, changed.

ovidiutirla commented on pkg/loadbalancer/experimental/lbmaps.go:
```diff
@@ -60,6 +67,100 @@ func newLBMapsConfig(dcfg *option.DaemonConfig) (cfg LBMapsConfig) {
+					expected := difflib.SplitLines(string(b))
+
+					var diff string
+					for range 10 {
```
I'm not quite following why we need this for loop?

joamaki replied:
Reconciliation is async so we don’t know exactly when it’s done and hence need to try few times. Same as with ”db cmp”. I might think about adding some support to reconciler to allow some sort of syncing for tests.

ovidiutirla replied:
Oh, I see now, we exit early if the len(diff) == 0. 

Wondering if it would make sense to extract the maxRetries and sleep as const? 
Also, I assume this pattern will be fairly common in the future, should we have some generic function for retries?

joamaki replied:
Yeah probably later I'll think about adding some sort of "eventually" thing that wraps the commands. It's not entirely trivial so didn't look into it yet.

Moved these to consts.

joamaki replied:
Here we go: https://github.com/cilium/hive/pull/22

ovidiutirla commented on pkg/loadbalancer/experimental/lbmaps.go:
```diff
@@ -242,6 +343,21 @@ func (r *BPFLBMaps) Stop(cell.HookContext) error {
 	return errors.Join(errs...)
 }
 
+// iteratemap iterates over a BPF map, yielding new keys and values. Similar to
```
```suggestion
// iterateMap iterates over a BPF map, yielding new keys and values. Similar to
```

ovidiutirla reviewed (approved):
LGTM 🚀

joamaki committed:
experimental: Add TestScript

Add scripttest runner

Signed-off-by: Jussi Maki <jussi@isovalent.com>

DamianSawicki commented on pkg/k8s/client/cell.go:
```diff
@@ -545,6 +587,86 @@ func FakeClientBuilder() ClientBuilderFunc {
+				// accepts we're good. We'll add to all since multiple trackers
+				// may accept (e.g. slim and kubernetes).
+
+				// err will get set to nil if any of the tracker methods succeed.
```
How about something like the following, which should be self-explanatory enough not to require extra comment?
```
success := false
checkErr := func(e error) {
  success := success || e == nil
}
(...)
if !success
...
```
And if we want to collect the errors, we can add `var errs []error`,`append(errs, e)` , and `err = errors.Join(errs)`.

joamaki replied:
We actually don't want to collect the errors as we do expect some of them to fail.

joamaki replied:
Ah actually useful to collect them as the error might be interesting, e.g. you might have 4 trackers rejecting it as unknown object and 1 rejecting it as known but malformed in some way. Changed this to collect all errors.

joamaki replied:
e.g. here's how "double add" looks like now:
```
        scripttest.go:212: FAIL: testdata/hostport.txtar:13: k8s add pod.yaml: none of the trackers of FakeClientset accepted *v1.Pod
            mcs: no kind is registered for the type v1.Pod in scheme "sigs.k8s.io/mcs-api/pkg/client/clientset/versioned/fake/register.go:30"
            kubernetes: no kind is registered for the type v1.Pod in scheme "k8s.io/client-go/kubernetes/fake/register.go:82"
            apiexit: no kind is registered for the type v1.Pod in scheme "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/fake/register.go:31"
            slim: pods "my-app-85f46c4bd9-nnk25" already exists
            cilium: no kind is registered for the type v1.Pod in scheme "github.com/cilium/cilium/pkg/k8s/client/clientset/versioned/fake/register.go:18"
```

earlier you probably would've just seen "no kind is registered" error and not the "already exists" that you wanted.

DamianSawicki replied:
Thanks! My main point was to use the `success` bool to avoid the need for comments (the current version has even two), but error collection and tracker iteration are also very nice changes (and your solution without bool has the benefit of avoiding the `bool+error` redundancy from my suggestion).

DamianSawicki commented on pkg/k8s/client/cell.go:
```diff
@@ -545,6 +587,86 @@ func FakeClientBuilder() ClientBuilderFunc {
+						err = e
+					}
+				}
+				switch action {
```
Nit: could we create a slice `[]testing.ObjectTracker` containing all of `fc.XXX.Tracker()`s before the switch and then just iterate over it for each case?

joamaki replied:
Good idea! Changed it so that the trackers are collected where the fake clients are created.. that way it's easy to extend.

DamianSawicki commented on pkg/k8s/testutils/resources.go:
```diff
@@ -0,0 +1,85 @@
+		},
+	}
+
+	Corev1APIResources = &metav1.APIResourceList{
```
Nit: rename `Corev1APIResources -> CoreV1APIResources` in line with `DiscoveryV1APIResources`, and similarly `Ciliumv2APIResources -> CiliumV2APIResources`. Super-nit: `DiscoveryV1beta1APIResources -> DiscoveryV1Beta1APIResources`.

joamaki replied:
Done

DamianSawicki commented on pkg/loadbalancer/experimental/lbmaps.go:
```diff
@@ -60,6 +107,109 @@ func newLBMapsConfig(dcfg *option.DaemonConfig) (cfg LBMapsConfig) {
+	}
+
+	var snapshot *mapSnapshots
+	errUsage := fmt.Errorf("%w: expected 'dump', 'snapshot' or 'restore'", script.ErrUsage)
```
`cmp` seems missing in this list.

joamaki replied:
added.

DamianSawicki commented on pkg/loadbalancer/experimental/lbmaps.go:
```diff
@@ -242,6 +392,21 @@ func (r *BPFLBMaps) Stop(cell.HookContext) error {
 }
 
+// iterateMap iterates over a BPF map, yielding new keys and values. Similar to
+// ebpf.IterateWithCallback but allocates new key & value instead of reusing.
```
What's the motivation? Is this lack of reusing a safeguard for the case when cb is non-blocking?

joamaki replied:
`ebpf.IterateWithCallback` reused the same allocated values which means we can't use it for snapshotting maps.

DamianSawicki commented on pkg/loadbalancer/experimental/lbmaps.go:
```diff
@@ -242,6 +392,21 @@ func (r *BPFLBMaps) Stop(cell.HookContext) error {
+// iterateMap iterates over a BPF map, yielding new keys and values. Similar to
+// ebpf.IterateWithCallback but allocates new key & value instead of reusing.
+func iterateMap(m *ebpf.Map, newPair func() (any, any), cb func(any, any)) error {
+	entries := m.Iterate()
```
The original `ebpf.IterateWithCallback` calls `m.lock.RLock()`. Why is it not needed here?

joamaki replied:
No need, it's safe to call `Iterate` concurrently. I don't know why `ebpf.Map` has a lock here.

joamaki replied:
Side note: this code really shouldn't use `pkg/ebpf` or `pkg/bpf`. We want to mostly get rid of them in the future and use `cilium/ebpf` directly. Might rewrite this pretty soon as there's very little in `pkg/ebpf` that it's building on.

DamianSawicki replied:
> I don't know why ebpf.Map has a lock here.

Perhaps the reason is to obtain a consistent state in case write happens concurrently?

joamaki replied:
The writes don't take the lock. It's probably just for guarding the `OpenOrCreate` call, which we don't do here.

DamianSawicki replied:
I see, thanks!

DamianSawicki commented on pkg/loadbalancer/experimental/reflector.go:
```diff
@@ -153,8 +154,6 @@ func runResourceReflector(ctx context.Context, p reflectorParams, initComplete f
 			}
 			txn := p.Writer.WriteTxn()
 			for _, ev := range buf {
-				ev.Done(nil)
```
This change looks unrelated to the PR. What's the rationale behind it?

joamaki replied:
Reason was that on sudden stopping some of the events were left unhandled causing the finalizer for the event to panic. Since we don't really care about leveraging the `Resource[T]` workqueue we can just mark them handled right away.

The plan for the reflector is to rewrite it to use client-go reflector directly and not use `Resource[T]` or an informer (no point having objects duplicated in a cache.Store), so this code will have a fairly short lifespan. I can move the change into its own commit if you'd like.

DamianSawicki replied:
> I can move the change into its own commit if you'd like.

That would be great, thanks!

joamaki replied:
Done

joamaki committed:
experimental: Always mark event as handled

Since we don't need to retry Resource[T] events in the reflector,
mark the events as handled right away. This way the reflector can
be stopped at any point with unhandled objects in the buffer without
causing the event finalizer to panic due to uncalled Done().

Signed-off-by: Jussi Maki <jussi@isovalent.com>

DamianSawicki commented on pkg/loadbalancer/experimental/script_test.go:
```diff
@@ -0,0 +1,110 @@
+	version.Force(testutils.DefaultVersion)
+
+	// pkg/k8s/endpoints.go uses this in ParseEndpointSlice*
+	option.Config.EnableK8sTerminatingEndpoint = true
```
Don't we want to defer restoring the original value?

joamaki replied:
Hmm restore it for whom? If some other test in this package depends on the value of this, then it should set it itself. Note that `option.Config` isn't actually initialized/populated, so it's in completely undefined state anyway...
Ugh I really want to get rid of these damn globals.

DamianSawicki replied:
Well, asking since not really sure, I was wondering that either now or in the future there could be some other test running afterwards? If not, let's leave it as it is for brevity.

DamianSawicki commented on pkg/loadbalancer/experimental/testdata/nodeport.txtar:
```diff
@@ -0,0 +1,271 @@
+status:
+  loadBalancer: {}
+
+-- endpointslice.yaml --
```
I see the old yamls contained 9 endpoints each and the new ones contain 4 endpoints each. Is this change intentional? Are there any other changes? From the commit description, this seems to be a simple copy-paste commit, so I'm not sure if anyone is reviewing the details.

joamaki replied:
I decided to simplify it. There wasn't really much point in having so many.

DamianSawicki reviewed (commented):
Thanks for all the changes and explanations!

dylandreimerink commented on pkg/k8s/client/cell.go:
```diff
@@ -517,10 +549,30 @@ func NewFakeClientset() (*FakeClientset, Clientset) {
 
+func toVersionInfo(rawVersion string) *versionapi.Info {
+	parts := strings.Split(rawVersion, ".")
+	return &versionapi.Info{Major: parts[0], Minor: parts[1]}
```
Is there a chance this is being called when running Cilium in a user environment? If so we should add guard here to prevent a panic on wrong input.

joamaki replied:
Only used by the fake client currently.

dylandreimerink commented on pkg/loadbalancer/experimental/benchmark/benchmark.go:
```diff
@@ -346,3 +352,243 @@ func ServicesAndSlices(testSize int) (svcs []*slim_corev1.Service, epSlices []*k
+	services chan resource.Event[*slim_corev1.Service],
+	pods chan resource.Event[*slim_corev1.Pod],
+	endpoints chan resource.Event[*k8s.Endpoints],
+	writer **experimental.Writer,
+	db **statedb.DB,
+	bo **experimental.BPFOps,
```
I understand this C style of doing things and why in this case. But its very un-golang like, I feel this deserves some inline comments for the uninitiated.

dylandreimerink reviewed (approved):
Got two nits, looks good to me otherwise
